File: branches/Medusa/Source/Fantasy.Jobs/Management/JobTemplatesService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Xml;
using Fantasy.Jobs.Properties;
using Fantasy.ServiceModel;

namespace Fantasy.Jobs.Management
{
    public interface IJobTemplatesService
    {
        JobTemplate GetJobTemplateByName(string name);
        JobTemplate GetJobTemplateByPath(string path);

        JobTemplate[] GetJobTemplates();
    }

    public class JobTemplatesService : AbstractService, IJobTemplatesService
    {

        private ILogger _logger;
        public ILogger Logger
        {
            get
            {
                if (_logger == null)
                {
                    _logger = (ILogger)this.Site.GetService(typeof(ILogger));
                }
                return _logger;
            }


        }

        private List<JobTemplate> _templates = new List<JobTemplate>();
        private List<JobTemplate> _invalidTemplates = new List<JobTemplate>();

        private int _sequence = 0;

        public override void InitializeService()
        {
            DirectoryInfo dir = new DirectoryInfo(JobManagerSettings.Default.JobTemplateDirectoryFullPath);
            if (dir.Exists)
            {
                LoadTemplates(dir);
            }
            base.InitializeService();
        }

        private void LoadTemplates(DirectoryInfo dir)
        {
            foreach(FileInfo fi in dir.GetFiles("*.jt"))
            {
                LoadTemplate(fi, false);
            }
            foreach (DirectoryInfo sub in dir.GetDirectories())
            {
                LoadTemplates(sub);
            }
        }

        private void LoadTemplate(FileInfo fi, bool replace)
        {
            JobTemplate template = new JobTemplate();
            template.id = this._sequence++;
            template.Location = fi.FullName;
            template.Valid = false;
            try
            {
                

                XmlDocument doc = new XmlDocument();

                string text = File.ReadAllText(fi.FullName);
                template.Content = text;
                doc.LoadXml(text);
                template.Name = doc.DocumentElement.GetAttribute("name");
                
                if (!String.IsNullOrEmpty(template.Name))
                {
                    JobTemplate old = (from t in this._templates where (StringComparer.OrdinalIgnoreCase.Compare(t.Name, template.Name) == 0) select t).SingleOrDefault();
                    if(old != null)
                    {
                        if(replace)
                        {
                            this._templates.Remove(old);
                        }
                        else
                        {
                            throw new ApplicationException(String.Format(Properties.Resources.DulplicateTemplateNamesText, old.Location, fi.FullName, old.Name));
                        }
                    }
                    template.Valid = true;
                    this._templates.Add(template);
                    Logger.LogMessage(LogCategories.Manager, Properties.Resources.SuccessToLoadTemplateText, fi.FullName, template.Name);
                }
                else
                {
                    template.Valid = true;
                    this._templates.Add(template);
                    
                    Logger.LogMessage(LogCategories.Manager, Properties.Resources.SuccessToLoadAnonymousTemplateText, fi.FullName);  
                }

            }
            catch(Exception ex)
            {
                Logger.LogWarning(LogCategories.Manager, ex, MessageImportance.High, Properties.Resources.FailToLoadTemplateText, fi.FullName);  
            }
            if (!template.Valid)
            {
                this._invalidTemplates.Add(template);
            }
        }




        #region IJobTemplatesService Members

        public JobTemplate GetJobTemplateByName(string name)
        {
            try
            {

                return (from template in this._templates where StringComparer.OrdinalIgnoreCase.Compare(template.Name, name) == 0 select template).Single();
            }
            catch (InvalidOperationException)
            {
                throw new JobException(String.Format(Properties.Resources.CannotFindTemplateByNameText, name)); 
            }

        }

        public JobTemplate GetJobTemplateByPath(string path)
        {
            Uri uri = new Uri(path);
            try
            {
                return (from template in this._templates where (new Uri(template.Location)) == uri select template).Single(); 
            }
            catch (InvalidOperationException)
            {
                throw new JobException(String.Format(Properties.Resources.CannotFindTemplateByPathText, path));
            }
        }

        #endregion


        public JobTemplate[] GetJobTemplates()
        {
            List<JobTemplate> rs = new List<JobTemplate>();
            rs.AddRange(this._templates);
            rs.AddRange(this._invalidTemplates);
            return rs.ToArray();
        }
    }
}

File: AutoConverted/Fantasy.Jobs/Management/JobTemplatesService.java
package Fantasy.Jobs.Management;

import Fantasy.Jobs.Properties.*;
import Fantasy.ServiceModel.*;

public class JobTemplatesService extends AbstractService implements IJobTemplatesService
{

	private ILogger _logger;
	public final ILogger getLogger()
	{
		if (_logger == null)
		{
			_logger = (ILogger)this.Site.GetService(ILogger.class);
		}
		return _logger;
	}



	private java.util.ArrayList<JobTemplate> _templates = new java.util.ArrayList<JobTemplate>();
	private java.util.ArrayList<JobTemplate> _invalidTemplates = new java.util.ArrayList<JobTemplate>();

	private int _sequence = 0;

	@Override
	public void InitializeService()
	{
		DirectoryInfo dir = new DirectoryInfo(JobManagerSettings.getDefault().getJobTemplateDirectoryFullPath());
		if (dir.Exists)
		{
			LoadTemplates(dir);
		}
		super.InitializeService();
	}

	private void LoadTemplates(DirectoryInfo dir)
	{
		for(FileInfo fi : dir.GetFiles("*.jt"))
		{
			LoadTemplate(fi, false);
		}
		for (DirectoryInfo sub : dir.GetDirectories())
		{
			LoadTemplates(sub);
		}
	}

	private void LoadTemplate(FileInfo fi, boolean replace)
	{
		JobTemplate template = new JobTemplate();
		template.setid(this._sequence++);
		template.setLocation(fi.FullName);
		template.setValid(false);
		try
		{


			XmlDocument doc = new XmlDocument();

			String text = File.ReadAllText(fi.FullName);
			template.setContent(text);
			doc.LoadXml(text);
			template.setName(doc.DocumentElement.GetAttribute("name"));

			if (!DotNetToJavaStringHelper.isNullOrEmpty(template.getName()))
			{
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
				JobTemplate old = (from t in this._templates where (StringComparer.OrdinalIgnoreCase.Compare(t.getName(), template.getName()) == 0) select t).SingleOrDefault();
				if(old != null)
				{
					if(replace)
					{
						this._templates.remove(old);
					}
					else
					{
						throw new ApplicationException(String.format(Properties.Resources.getDulplicateTemplateNamesText(), old.getLocation(), fi.FullName, old.getName()));
					}
				}
				template.setValid(true);
				this._templates.add(template);
				getLogger().LogMessage(LogCategories.getManager(), Properties.Resources.getSuccessToLoadTemplateText(), fi.FullName, template.getName());
			}
			else
			{
				template.setValid(true);
				this._templates.add(template);

				getLogger().LogMessage(LogCategories.getManager(), Properties.Resources.getSuccessToLoadAnonymousTemplateText(), fi.FullName);
			}

		}
		catch(RuntimeException ex)
		{
			getLogger().LogWarning(LogCategories.getManager(), ex, MessageImportance.High, Properties.Resources.getFailToLoadTemplateText(), fi.FullName);
		}
		if (!template.getValid())
		{
			this._invalidTemplates.add(template);
		}
	}




//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#region IJobTemplatesService Members

	public final JobTemplate GetJobTemplateByName(String name)
	{
		try
		{

//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
			return (from template in this._templates where StringComparer.OrdinalIgnoreCase.Compare(template.getName(), name) == 0 select template).Single();
		}
		catch (InvalidOperationException e)
		{
			throw new JobException(String.format(Properties.Resources.getCannotFindTemplateByNameText(), name));
		}

	}

	public final JobTemplate GetJobTemplateByPath(String path)
	{
		Uri uri = new Uri(path);
		try
		{
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
			return (from template in this._templates where (new Uri(template.getLocation())) == uri select template).Single();
		}
		catch (InvalidOperationException e)
		{
			throw new JobException(String.format(Properties.Resources.getCannotFindTemplateByPathText(), path));
		}
	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#endregion


	public final JobTemplate[] GetJobTemplates()
	{
		java.util.ArrayList<JobTemplate> rs = new java.util.ArrayList<JobTemplate>();
		rs.addAll(this._templates);
		rs.addAll(this._invalidTemplates);
		return rs.toArray(new JobTemplate[]{});
	}
}